### PR TITLE
Single app server

### LINKED ISSUE #383: Create single-application Oak server
Implement a new binary called `oak_server` or similar, which is effectively a slimmed down version of the Oak Manager that loads a single Oak Application provided as a command line argument (i.e. the argument points to a file containing a textual or binary representation of the Oak Application).

`oak_server` creates a single enclave with an instance of the Oak Runtime and loads the provided Application there.

`oak_server` does not expose an untrusted gRPC interface, though the Oak Application that it runs may have a gRPC node that exposes a trusted gRPC interface (used by clients to interact with the Oak Application).

It is not possible to create multiple enclaves or to change the Oak Application that is running; instead this would require restarting the `oak_server` instance.

Related to #382 (but not necessarily blocking in either direction).

### PR DESCRIPTION
**Pull request by ipetr0v**
This change:
- Adds a single application server
- Adds single app docker images in all examples
- Adds an explicit gRPC port in gRPC pseudo-node

Fixes #383

### CODE REVIEW

File: examples/hello_world/client/BUILD [hunk 1]
```diff
@@ -25,13 +26,31 @@ cc_binary(
...
+        "//oak/common:utils",
         "@com_github_grpc_grpc//:grpc++",
         "@com_google_absl//absl/flags:flag",
         "@com_google_absl//absl/flags:parse",
         "@com_google_absl//absl/memory",
         "@com_google_asylo//asylo/util:logging",
     ],
 )
+
+container_image(
```
Comment (tiziano88) [inline examples/hello_world/client/BUILD]: Can you add a TODO(#xxx) here and create an issue to encapsulate this in a `oak_application_server_docker` rule or something like that?

File: oak/proto/manager.proto [hunk 2]
```diff
@@ -33,6 +33,8 @@ message ApplicationConfiguration {
   // Indication of what configuration the initial Node should run.  Must identify a
   // NodeConfiguration entry that holds a WebAssemblyConfiguration object.
   string initial_node = 2;
+  // Port number used by the gRPC pseudo-node. If not set, port will be chosen automatically.
```
Comment (tiziano88) [inline oak/proto/manager.proto]:
```suggestion
  // Port number used by the gRPC pseudo-node. If zero or unset, port will be chosen automatically, and returned as part of the `CreateApplicationResponse` message.
```

File: oak/server/asylo/BUILD [hunk 3]
```diff
@@ -21,7 +21,11 @@ load("@linux_sgx//:sgx_sdk.bzl", "sgx")
 load("@rules_cc//cc:defs.bzl", "cc_library")
 
 package(
-    default_visibility = ["//oak/server:__subpackages__"],
+    #default_visibility = ["//visibility:public"],
```
Comment (tiziano88) [inline oak/server/asylo/BUILD]: Fix or remove.

File: oak/server/asylo/asylo_oak_app_main.cc [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#include "oak/common/app_config.h"
+#include "oak/common/utils.h"
+#include "oak/proto/manager.grpc.pb.h"
+#include "oak/server/asylo/asylo_oak_manager.h"
+
+ABSL_FLAG(std::string, enclave_path, "", "Path of the enclave to load");
+ABSL_FLAG(std::string, module, "", "File containing the compiled WebAssembly module");
+ABSL_FLAG(std::string, storage_address, "127.0.0.1:7867",
+          "Address ot the storage provider to connect to");
+ABSL_FLAG(int, grpc_port, 8080, "Port to listen on");
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_app_main.cc]:
```suggestion
ABSL_FLAG(int, grpc_port, 8080, "Port for the Application to listen on");
```

File: oak/server/asylo/asylo_oak_app_main.cc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  // We install an explicit SIGINT handler, as for some reason the default one
+  // does not seem to work.
+  std::signal(SIGINT, sigint_handler);
+
+  // Create manager instance.
+  std::unique_ptr<oak::AsyloOakManager> manager =
+      absl::make_unique<oak::AsyloOakManager>(absl::GetFlag(FLAGS_enclave_path));
+
+  // Load the Oak Module to execute. This needs to be compiled from Rust to WebAssembly separately.
+  std::string module_bytes = oak::utils::read_file(absl::GetFlag(FLAGS_module));
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_app_main.cc]: As discussed, eventually this should read a pre-packaged application configuration proto. Could you add a TODO here, and create two issues, one to create the actual "packer" (or whatever you want to call it), plus optionally a bazel rule that creates it based on wasm files, and then change this code to just read it from the proto file.

File: oak/server/asylo/asylo_oak_app_main.cc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#include "include/grpcpp/server.h"
+#include "include/grpcpp/server_builder.h"
+#include "oak/common/app_config.h"
+#include "oak/common/utils.h"
+#include "oak/proto/manager.grpc.pb.h"
+#include "oak/server/asylo/asylo_oak_manager.h"
+
+ABSL_FLAG(std::string, enclave_path, "", "Path of the enclave to load");
+ABSL_FLAG(std::string, module, "", "File containing the compiled WebAssembly module");
+ABSL_FLAG(std::string, storage_address, "127.0.0.1:7867",
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_app_main.cc]: As an aside, should we use more IPv6 flags? I wonder whether IPv6 is even supported on GCP though, could you check, and then create an issue to use IPv6 everywhere in our code?

File: oak/server/asylo/asylo_oak_manager_service.cc [hunk 7]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2018 The Project Oak Authors
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_manager_service.cc]:
```suggestion
 * Copyright 2019 The Project Oak Authors
```

File: oak/server/asylo/asylo_oak_manager.h [hunk 8]
```diff
@@ -28,23 +28,20 @@
 
 namespace oak {
 
-class AsyloOakManager final : public Manager::Service {
+class AsyloOakManager {
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_manager.h]: Maybe we should call this `AsyloOakLoader` or something like that, and keep the `Manager` name for the gRPC version only, WDYT?

File: oak/server/asylo/asylo_oak_app_main.cc (hunk 5 above)
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_app_main.cc]: Done
https://github.com/project-oak/oak/issues/412
https://github.com/project-oak/oak/issues/413

File: oak/server/asylo/asylo_oak_app_main.cc (hunk 6 above)
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_app_main.cc]: If I understand correctly, we only can create external IPv6 addresses, that can be mapped to the internal IPv4 ones:
https://cloud.google.com/load-balancing/docs/ipv6

File: oak/server/asylo/asylo_oak_app_main.cc [hunk 9]
```diff
@@ -51,18 +51,21 @@ int main(int argc, char* argv[]) {
   std::signal(SIGINT, sigint_handler);
 
   // Create manager instance.
-  std::unique_ptr<oak::AsyloOakManager> manager =
-      absl::make_unique<oak::AsyloOakManager>(absl::GetFlag(FLAGS_enclave_path));
+  std::unique_ptr<oak::AsyloOakLoader> loader =
+      absl::make_unique<oak::AsyloOakLoader>(absl::GetFlag(FLAGS_enclave_path));
 
   // Load the Oak Module to execute. This needs to be compiled from Rust to WebAssembly separately.
+  // TODO: Use a `Config` file as an input:
```
Comment (tiziano88) [inline oak/server/asylo/asylo_oak_app_main.cc]: `TODO(#412, #413):` (and elsewhere)

File: examples/hello_world/client/BUILD (hunk 1 above)
Comment (tiziano88) [reply examples/hello_world/client/BUILD]: Which issue is that? Could you use the format `TODO(#xxx)` where `xxx` is the issue number in GItHub, when possible? So it's easier to see when a PR fixes an issue.

File: oak/server/asylo/asylo_oak_loader.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+                              const oak::ApplicationConfiguration& application_configuration);
+
+  asylo::StatusOr<InitializeOutput> GetEnclaveOutput(const std::string& application_id);
+
+  std::string NewApplicationId();
+
+  void DestroyEnclave(const std::string& node_id);
+
+  asylo::EnclaveManager* enclave_manager_;
+  std::unique_ptr<asylo::SimLoader> enclave_loader_;
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_loader.h]: Done

File: oak/server/asylo/asylo_oak_app_main.cc (hunk 9 above)
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_app_main.cc]: Done

File: oak/server/asylo/asylo_oak_main.cc [hunk 11]
```diff
@@ -30,7 +30,7 @@
 #include "include/grpcpp/server_builder.h"
 
 ABSL_FLAG(std::string, enclave_path, "", "Path of the enclave to load");
-ABSL_FLAG(int, grpc_port, 8888, "Port to listen on");
+ABSL_FLAG(int, grpc_port, 8888, "Port for the Application to listen on");
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_main.cc]: Nothing changed in the original case.
And in the new case (single app) server loads application on startup, and the only active gRPC port is the application port.

File: oak/server/asylo/asylo_oak_manager.h [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 The Project Oak Authors
+ * Copyright 2019 The Project Oak Authors
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_manager.h]: Fixed in https://github.com/project-oak/oak/pull/423

File: oak/server/asylo/asylo_oak_manager.h [hunk 13]
```diff
@@ -41,23 +43,9 @@ class AsyloOakManager final : public Manager::Service {
...
-  std::string NewApplicationId();
-
-  void DestroyEnclave(const std::string& node_id);
-
-  asylo::EnclaveManager* enclave_manager_;
-  std::unique_ptr<asylo::SimLoader> enclave_loader_;
-  std::string enclave_path_;
-
-  uint64_t application_id_;
+  oak::AsyloOakLoader manager_;
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_manager.h]: Fixed in https://github.com/project-oak/oak/pull/423

File: oak/common/BUILD [hunk 14]
```diff
@@ -90,3 +90,13 @@ cc_library(
         "@com_google_absl//absl/strings",
     ],
 )
+
+cc_library(
+    name = "utils",
+    hdrs = ["utils.h"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "@com_github_grpc_grpc//:grpc++",
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/common/BUILD]: Fixed in https://github.com/project-oak/oak/pull/423

File: oak/common/app_config.cc [hunk 15]
```diff
@@ -58,6 +58,10 @@ void AddStorageToConfig(ApplicationConfiguration* config, const std::string& sto
   storage->set_address(storage_address);
 }
 
+void AddGrpcPortToConfig(ApplicationConfiguration* config, const int16_t grpc_port) {
+  config->set_grpc_port(grpc_port);
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/common/app_config.cc]: Fixed in https://github.com/project-oak/oak/pull/423

File: oak/server/asylo/asylo_oak_main.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/asylo/asylo_oak_main.cc]: (Sorry, I don't think I was very clear in the first comment)

This changes the help text for the original case (the Asylo-based generic server):
```console
% ./bazel-bin/oak/server/asylo/oak --grpc_port=8888 --helpfull
[snip]
    -grpc_port (Port for the Application to listen on); default: 8888;
      currently: 8888
%
```

So the general Oak server (this **a**pplication) listens on :8888, and when a client starts a new Oak **A**pplication, that will listen on a new port &ndash; not the one specified by `--grpc_port` &ndash; and that makes the help message a bit misleading.

[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak/server/asylo/asylo_oak_main.cc]: The comment was inserted in a wrong file.
Fixed in https://github.com/project-oak/oak/pull/435

**Pull request merged**